keeper/src/main/scala/zio/keeper/hyparview/protocols.scala

```diff
@@ -5,35 +5,86 @@ import zio.stm._
 import zio.keeper._
 import zio.stream.ZStream
 import zio.keeper.transport.{ Connection, Protocol }
+import zio.keeper.transport.Transport
 
 object protocols {
 
-  def hyparview[R <: HyParViewConfig with Views with TRandom, E](
-    con: Connection[R, E, Message, Message],
-    peerMessages: Enqueue[(NodeAddress, Message.PeerMessage)]
+  def runInitial[R <: HyParViewConfig with Views with TRandom, E](
+    con: Connection[R, E, Message, Message]
   ): ZIO[R, E, Unit] =
     ZManaged.switchable[R, E, Boolean => UIO[Unit]].use { switch =>
       val protocol =
         initialProtocol
           .contM[R, E, Message, Message, Any] {
             case Some(remoteAddr) =>
               switch(inActiveView(remoteAddr, con.send)).map { setKeepRef =>
-                Right(activeProtocol(remoteAddr, peerMessages.contramap((remoteAddr, _))).onEnd(setKeepRef).unit)
+                Right(activeProtocol(remoteAddr).onEnd(setKeepRef).unit)
               }
             case None =>
               ZIO.succeedNow(Left(()))
           }
-      Protocol.run(con, protocol).unit
+      protocol.run(con).unit
     }
 
+  def runActive[R <: HyParViewConfig with Views with TRandom, E](
+    remoteAddress: NodeAddress,
+    con: Connection[R, E, Message, Message]
+  ): ZIO[R, E, Unit] =
+    inActiveView(remoteAddress, con.send).use { setKeepRef =>
+      activeProtocol(remoteAddress).onEnd(setKeepRef).run(con).unit
+    }
+
+  // start protocol with remote node
+  def connectRemote(
+    addr: NodeAddress,
+    initialMessage: Message
+  ): ZIO[Transport with HyParViewConfig with Views with TRandom, Error, Unit] = {
+    val initial = Transport.connect(addr).map(_.withCodec[Message]()).use { con =>
+      con.send(initialMessage) *> protocols.runInitial(con)
+    }
+    val active = Transport.connect(addr).map(_.withCodec[Message]()).use { con =>
+      con.send(initialMessage) *> protocols.runActive(addr, con)
+    }
+    val nothing = ZIO.unit
+
+    initialMessage match {
+      case _: Message.ShuffleReply =>
+        initial
+      case Message.Neighbor(_, true) =>
+        active
+      case Message.Neighbor(_, false) =>
+        Transport
+          .connect(addr)
+          .map(_.withCodec[Message]())
+          .use { con =>
+            for {
+              _     <- con.send(initialMessage)
+              reply <- con.receive.runHead
+              _ <- reply match {
+                    case Some(Message.NeighborAccept) =>
+                      protocols.runActive(addr, con)
+                    case _ =>
+                      Views.doNeighbor.commit
+                  }
+            } yield ()
+          }
+          .onError(_ => Views.doNeighbor.commit)
+      case _: Message.Join =>
+        active
+      case _: Message.ForwardJoinReply =>
+        active
+      case _ =>
```

Comment:
We could introduce another `Message` subtype to remove this case, but I don't think it's worth the complexity.